File: mj/[PGS]다단계 칫솔 판매.cpp
#include <string>
#include <vector>
#include <unordered_map>

using namespace std;
unordered_map<string, string> refer_info;   // key: enroll ,value:referral
unordered_map<string, int> money_info;      // key: seller ,value:amount

// 수익분배 계산 함수. 재귀로 구현
void moneyget(string person,int money){    
    if(person.compare("-")==0) return;  // 추천인이 없으면 return
    int togive = money * 0.1;       // 추천인에게 줘야하는 수익의 10% 금액
    money_info[person] += (money - togive); // 추천인에게 줄 것 제외하고 현재 판매원의 몫
    if(togive<1) return;    // 추천인에게 안 줘도 되면 return 
    moneyget(refer_info[person], togive);
}

vector<int> solution(vector<string> enroll, vector<string> referral, vector<string> seller, vector<int> amount) {
    vector<int> answer;

    // 두 uo_m 준비
    for(int i = 0 ; i<enroll.size();i++){
        refer_info.insert(pair<string, string>(enroll[i],referral[i]));
        money_info.insert(pair<string, int>(enroll[i],0));
    }
    
    // 판매실적을 낸 판매원들을 기준으로 수익분배 계산
    for(int i =0; i<seller.size();i++){
        int salesMoney = amount[i] * 100;        // 칫솔개수 *100원
        moneyget(seller[i] ,salesMoney); 
    }
    
    for(int i = 0; i< enroll.size();i++)
        answer.push_back(money_info[enroll[i]]);
    
    return answer;
}

File: ms/[PGS] 다단계 칫솔 판매.py
#### 테케는 다 맞는데 채점해보면 다 틀리는 이유는????

# enroll = 각 판매원의 이름(center 이름은 없음)
# referral = 각 판매원의 추천인
# seller = 판매량에 대한 판매원의 이름
# amount = 판매량
def solution(enroll, referral, seller, amount):
    amount = [i * 100 for i in amount]
    answer = [0 for i in enroll]

    # enroll의 추천인 인덱스 배열 rindex -> 예시 [0, 0, 1, 2, 1, 1, 5, 2]
    rindex = []
    for i in range(len(enroll)):
        if referral[i] != '-':
            rindex.append(enroll.index(referral[i]))
        else:
            rindex.append(0)

    """
    판매원 young의 경우
    반복 1 -> staff = 7, money = 1200, boss = 2, answer[7] += 1080
    반복 2 -> money = 120, staff = 2, boss = 1, answer[2] += 108
    반복 3 -> money = 12, staff = 1, boss = 0(break문)
    이런 식으로 풀었음!
    """
    for i in range(len(seller)):
        staff = enroll.index(seller[i])
        money = amount[i]

        while True:
            answer[staff] += (money - money // 10)
            boss = rindex[staff]

            if boss == 0:
                break
            
            money = money // 10
            staff = boss
    
    return answer
